lib/matplotlib/testing/determinism.py

```diff
@@ -0,0 +1,136 @@
+"""
+Provides utilities to test output reproducibility.
+"""
+
+import io
+import os
+import re
+
+from matplotlib import pyplot as plt
+
+
+def _test_determinism_save(filename, objects='mhi', format="pdf"):
+    # save current value of SOURCE_DATE_EPOCH and set it
+    # to a constant value, so that time difference is not
+    # taken into account
+    sde = os.environ.pop('SOURCE_DATE_EPOCH', None)
+    os.environ['SOURCE_DATE_EPOCH'] = "946684800"
+
+    fig = plt.figure()
+
+    if 'm' in objects:
+        # use different markers...
+        ax1 = fig.add_subplot(1, 6, 1)
+        x = range(10)
+        ax1.plot(x, [1] * 10, marker=u'D')
+        ax1.plot(x, [2] * 10, marker=u'x')
+        ax1.plot(x, [3] * 10, marker=u'^')
+        ax1.plot(x, [4] * 10, marker=u'H')
+        ax1.plot(x, [5] * 10, marker=u'v')
+
+    if 'h' in objects:
+        # also use different hatch patterns
+        ax2 = fig.add_subplot(1, 6, 2)
+        bars = ax2.bar(range(1, 5), range(1, 5)) + \
+            ax2.bar(range(1, 5), [6] * 4, bottom=range(1, 5))
+        ax2.set_xticks([1.5, 2.5, 3.5, 4.5])
+
+        patterns = ('-', '+', 'x', '\\', '*', 'o', 'O', '.')
+        for bar, pattern in zip(bars, patterns):
+            bar.set_hatch(pattern)
+
+    if 'i' in objects:
+        # also use different images
+        A = [[1, 2, 3], [2, 3, 1], [3, 1, 2]]
+        fig.add_subplot(1, 6, 3).imshow(A, interpolation='nearest')
+        A = [[1, 3, 2], [1, 2, 3], [3, 1, 2]]
+        fig.add_subplot(1, 6, 4).imshow(A, interpolation='bilinear')
+        A = [[2, 3, 1], [1, 2, 3], [2, 1, 3]]
+        fig.add_subplot(1, 6, 5).imshow(A, interpolation='bicubic')
+
+    x = range(5)
+    fig.add_subplot(1, 6, 6).plot(x, x)
+
+    fig.savefig(filename, format=format)
```

Comment:
Since they're running in external processes, I don't think `BytesIO` would be the right approach. What might work is to output to `stdout` and read from there with `check_output`. I can open a PR to show how this might work.